addons/udes_report/models/stock_export.py

```diff
@@ -62,79 +62,99 @@ def run_stock_file_export(self, context=None, send_file_via='user'):
+                    for loc, _quants in groupby(quant_union.sorted(key=by_location),
+                                                key=by_location)}
+
+        locations = self.included_locations - self.excluded_locations
+        if not locations:
+            raise UserError(_("The specified list of Stock Locations is empty."))
+
+        all_quants = Quant.search([('location_id', 'child_of', locations.ids)])
 
         quants_by_prod_by_loc = OrderedDict()
         prod_summary = defaultdict(lambda: {'n_pkgs': 0, 'qty': 0})
 
-        for prod, _quants in groupby(all_quants.sorted(key=by_prod),
-                                     key=by_prod):
+        for prod, _quants in groupby(all_quants.sorted(key=by_prod), key=by_prod):
             quant_prods = Product.browse(prod)
             quants_u = Quant.union(*_quants)
             quants_by_prod_by_loc[quant_prods] = get_quants_by_loc(quants_u)
 
-        def get_prod_qty(quants):
-            return sum(Quant.union(*quants).mapped('quantity'))
-
-        _logger.info(_('Creating Excel file'))
-        file_name = 'warehouse_stock_{TIMESTAMP}.xls'.format(
-            TIMESTAMP=datetime.now())
-        wb = xlwt.Workbook()
-
-        # 1/2) 'Stock File': one entry for each (prod, package) pair
+        # 1/2) 'Stock File': one entry for each prod, package and lot (if applicable)
+        stock_file_data = {"name": 'Stock File'}
+        # Determine if lots/serials should be shown on the report
+        show_tracking = bool(Product.search_count([("tracking", "!=", "none")]))
 
-        stock_sheet = _create_sheet(wb, 'Stock File', ['Part Number',
-                                                       'Location',
-                                                       'Package',
-                                                       'Quantity'])
+        headers = ['Part Number', 'Location', 'Package', 'Quantity']
+        if show_tracking:
+            headers.insert(3, 'Lot/Serial Number')
+        stock_file_data["column_titles"] = headers
 
-        row = 0
+        rows = []
         for prod, quants_by_location in quants_by_prod_by_loc.items():
             for location, quants in quants_by_location.items():
-                for pkg_name, grouped_quants in groupby(quants,
+                for pkg_name, grouped_quants in groupby(quants.sorted(key=by_package_name),
                                                         key=by_package_name):
-                    row += 1
-                    qty = get_prod_qty(grouped_quants)
-
-                    stock_sheet.write(row, 0, prod.default_code)
-                    stock_sheet.write(row, 1, location.display_name)
-                    stock_sheet.write(row, 2, pkg_name)
-                    stock_sheet.write(row, 3, qty)
-
-                    prod_summary[prod]['qty'] += qty
+                    # If lots are not present then this will not change the rows
+                    for lot_name, lot_grouped_quants in groupby(grouped_quants,
```

Comment:
We can now replace `groupby(grouped_quants, key=by_lot_name)` by `grouped_quants.grupby(by_lot_name)` and `lot_grouped_quants` will be a recordset already, so we don't need `Quant.union` at line 133, and we can do something like: `lot_grouped_quants.total_quantity()` which does the sum.

Reply:
Changed, looks much better